bundles/org.corpus_tools.hexatomic.graph/src/main/java/org/corpus_tools/hexatomic/graph/internal/SaltGraphLayout.java

```diff
@@ -0,0 +1,425 @@
+              // previous occupancy)
+              occupiedByText.clear();
+            }
+          }
+
+          // update the occupied region to include the new sequences
+          for (DataSourceSequence<?> s : sequences) {
+            if (s.getDataSource() instanceof STextualDS) {
+              BitSet occupied = occupiedByText.get(s.getDataSource());
+              if (occupied == null) {
+                occupied = new BitSet();
+                occupiedByText.put((STextualDS) s.getDataSource(), occupied);
+              }
+              occupied.set(s.getStart().intValue(), s.getEnd().intValue());
+            }
+          }
+
+          // remember the generated rank and sub-rank
+          mergedRanks.put(internal, new RankSubrank(rank, subrank));
+        }
+      }
+    }
+
+    // re-create the ranks map but include the merged sub-ranks
+    rankForNode.clear();
+    int flattenedRank = 0;
+    RankSubrank previousRank = null;
+    // Because mergedRanks is a linked hash map and the insertion order was by the original rank,
+    // the iteration will be sorted by the original rank value (smallest rank first)
+    for (Map.Entry<InternalNode, RankSubrank> e : mergedRanks.entrySet()) {
+      InternalNode internal = e.getKey();
+      RankSubrank currentRank = e.getValue();
+
+      if (previousRank != null && !currentRank.equals(previousRank)) {
+        // add a new global (or flattened) rank
+        flattenedRank++;
+      }
+
+      rankForNode.put(internal, flattenedRank);
+
+      previousRank = currentRank;
+    }
+
+    fireProgressEvent(2, 4);
+
+    // 3. layout tokens and put them one rank below the number of ranks
+    // TODO: check if any pointing relations are present and use +1 if not
+    layoutTokenOrder(tokens, boundsX, boundsY, flattenedRank + 2);
+    fireProgressEvent(3, 4);
+
+    // 4. assign position based on (sub-) rank and the covered tokens
+    for (Map.Entry<InternalNode, Integer> e : rankForNode.entrySet()) {
+      int rank = e.getValue();
+      InternalNode node = e.getKey();
+      SNode saltNode = this.nodes.get(node);
+      // Get covered tokens to get the x position
+      if (saltNode.getGraph() instanceof SDocumentGraph) {
+        SDocumentGraph graph = (SDocumentGraph) saltNode.getGraph();
+        List<SToken> overlappedTokens = graph.getOverlappedTokens(saltNode);
+        Range<Double> range = getTokenRange(overlappedTokens);
+
+        double x = 0.0;
+        if (range.hasLowerBound()) {
+          if (range.hasUpperBound()) {
+            // position at the center
+            x = range.lowerEndpoint() + (range.upperEndpoint() - range.lowerEndpoint()) / 2.0;
+          } else {
+            // position left-aligned
+            x = range.lowerEndpoint();
+          }
+        }
+        node.setInternalLocation(boundsX + x,
+            boundsY + (rank * (this.maxNodeHeight * this.percentMargin)));
+
+      }
+    }
+
```

Comment:
```suggestion

fireProgressEvent(4, 4);

```

Fire notice that last step is complete.